### PR TITLE
setDT generates shallow copy earlier to avoid interfering with attributes of co-bound tables

### LINKED ISSUE #4784: setDT modifies the class of the origin data.frame
Take this code:
```
> d1 <- data.frame(a=c(1,2,3,4), b=c(2,3,4,5))
> d2 <- d1
> setDT(d2)  
```
At this point d2 is a shallow copy of d1, meaning they are not the same R object (although they point to the same columns).
`d2`'s class is data.table, as expected, _but so is `d1`_
```
> class(d2)
[1] "data.table" "data.frame"
> class(d1)
[1] "data.table" "data.frame"
```
Unlike #4783 , I think this is an easy fix. In two places in `setDT`'s source you see the pair of lines -
```
setattr(x, "class", .resetclass(x, "data.frame"))
setalloccol(x)
```
`setalloccol` is what initiates the shallow copy. If the order of the rows were reversed, the class of only the current shallow copy (`d2` in the example above) would be modified, as expected.

`#` `Output of sessionInfo()`
R version 3.6.1 (2019-07-05)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19041)
Matrix products: default

### PR DESCRIPTION
**Pull request by MichaelChirico**
Closes #4784. Original author: @OfekShilon.

This is #6470, reborn (again). I tried resolving the merge conflict on that PR but I was rejected trying to push to Ofek's fork:

```
 ! [remote rejected]   fix4784 -> fix4784 (refusing to allow a GitHub App to create or update workflow `.github/workflows/performance-tests.yml` without `workflows` permission)
error: failed to push some refs to 'https://github.com/OfekShilon/data.table.git'
```

Just FYI to @OfekShilon -- now that you're a project member, should the need arise, making branches on this repo going forward will avoid such issues.

### CODE REVIEW

**Comment by OfekShilon** [comment]
Thank you @MichaelChirico. I feel I should say - not only I don't have access to my original source repo, but I'm also currently not working with data.table and do not expect to be contributing in the foreseeable future. If you prefer to exclude me from the project members I completely understand.

**Comment by MichaelChirico** [comment]
No worries @OfekShilon, you had mentioned similar elsewhere (can't see it right away). My comment is as much of a note to self/fellow readers as to you :)

Right now we don't have any policy for removing ppl from 'Maintainers', so the easier thing is to leave you in there and hope we'll see you back eventually! Thanks for all your contributions over the years :pray:

**Comment by MichaelChirico** [comment]
@ben-schwen do you think there's any test we should add for your 2nd suggested change (running `setalloccol()` before `setattr()` for row.names?

**Comment by MichaelChirico** [comment]
Answering my own question, run at 899e49b5c185b64224030c5e2bcc5ab51992bdbd

```
d1=data.frame(a=1, row.names='b')
d1
#   a
# b 1
d2=d1
setDT(d2)
d1
#   a
# 1 1
```

**Comment by MichaelChirico** [comment]
@ben-schwen based on the test failures I _think_ this approach is the right way to go -- rather than relying on `setalloccol()` doing a shallow copy for us (which IINM it only does in certain cases), we are better of doing it directly. I don't _think_ there's any issue with doing `x = .shallow(x)` here, right?

File: NEWS.md [hunk 1]
```diff
@@ -111,6 +111,8 @@ rowwiseDT(
 
 13. Restore some join operations on `x` and `i` (e.g. an anti-join `x[!i]`)  where `i` is an extended data.frame, but not a data.table (e.g. a `tbl`), [#6501](https://github.com/Rdatatable/data.table/issues/6501). Thanks @MichaelChirico for the report and PR.
 
+14. `setDT()` no longer modifies the class of other names bound to the origin data.frame - e.g., argument DFs names at a caller to a function which uses setDT. Cf [#4784](https://github.com/Rdatatable/data.table/issues/4784). Thanks @OfekShilon for the report and fix.
```
Comment (MichaelChirico) [inline NEWS.md:114]:
```suggestion
14. `setDT()` no longer modifies the class of other names bound to the origin data.frame, e.g., in `DF1 <- data.frame(a=1); DF2 <- DF1; setDT(DF2)`, `DF1`'s class will not change. [#4784](https://github.com/Rdatatable/data.table/issues/4784). Thanks @OfekShilon for the report and fix.
```

File: R/data.table.R [hunk 2]
```diff
@@ -2922,12 +2922,18 @@ setDT = function(x, keep.rownames=FALSE, key=NULL, check.names=FALSE) {
...
       }
     }
+
+    # Done to avoid affecting other copies of x when we setattr() below (#4784)
+    x = .shallow(x)
+
     rn = if (!identical(keep.rownames, FALSE)) rownames(x) else NULL
     setattr(x, "row.names", .set_row_names(nrow(x)))
     if (check.names) setattr(x, "names", make.names(names(x), unique=TRUE))
+
```
Comment (MichaelChirico) [inline R/data.table.R:2932]:
```suggestion
```

File: R/data.table.R [hunk 3]
```diff
@@ -2922,12 +2922,18 @@ setDT = function(x, keep.rownames=FALSE, key=NULL, check.names=FALSE) {
...
     # fix for #1078 and #1128, see .resetclass() for explanation.
     setattr(x, "class", .resetclass(x, 'data.frame'))
     setalloccol(x)
+
```
Comment (MichaelChirico) [inline R/data.table.R:2936]:
```suggestion
```

**Review by Anirban166 — APPROVED** [review]

**Comment by MichaelChirico** [comment]
@Anirban166 aside: I think there's too much compression of the image for atime still:

image

Possibly because the new diagnostic text below is taking up the vertical real estate?

**Comment by Anirban166** [comment]
You're right, the height looks shrunk as well to me now (Perhaps [this line](https://github.com/tdhock/atime/blob/4fbe57b11f55f1e8e96a940a465c42954e19b4db/R/test.R#L171)?)

(@tdhock before you push `atime` to CRAN again like you mentioned [here](https://github.com/Anirban166/Autocomment-atime-results/issues/37#issuecomment-2380315024), please have a look at this too)

> Possibly because the new diagnostic text below is taking up the vertical real estate?

I don't see how that affects the image - All I'm doing towards the end of my GHA or in the last step is just pushing everything into a markdown file, which should separately render both the image (I do not tweak it, or say specify a manual height attribute) and the text below it

**Comment by Anirban166** [comment]
You're right, the height looks shrunk as well to me now (Perhaps something to do with [this line](https://github.com/tdhock/atime/blob/4fbe57b11f55f1e8e96a940a465c42954e19b4db/R/test.R#L171)?)

(@tdhock before you push `atime` to CRAN again like you mentioned [here](https://github.com/Anirban166/Autocomment-atime-results/issues/37#issuecomment-2380315024), please have a look at this too)

> Possibly because the new diagnostic text below is taking up the vertical real estate?

I don't see how that affects the image - All I'm doing towards the end of my GHA or in the last step is just pushing everything into a markdown file, which should separately render both the image (I do not tweak it, or say specify a manual height attribute) and the text below it

**Comment by Anirban166** [comment]
You're right, the height looks shrunk as well to me now (Perhaps something to do with [this line](https://github.com/tdhock/atime/blob/4fbe57b11f55f1e8e96a940a465c42954e19b4db/R/test.R#L171)?)

(@tdhock before you push `atime` to CRAN again but at a later point like you mentioned [here](https://github.com/Anirban166/Autocomment-atime-results/issues/37#issuecomment-2380315024), please have a look at this too)

> Possibly because the new diagnostic text below is taking up the vertical real estate?

I don't see how that affects the image - All I'm doing towards the end of my GHA or in the last step is just pushing everything into a markdown file, which should separately render both the image (I do not tweak it, or say specify a manual height attribute) and the text below it

**Comment by Anirban166** [comment]
(AFAIK there is no strict vertical limit to the content for a GitHub comment (just 65536 characters at max), and [`cml`](https://github.com/iterative/cml) follows the same rules of markdown)

**Comment by Anirban166** [comment]
(AFAIK there is also no strict vertical limit to the content for a GitHub comment (just 65536 characters at max), and [cml](https://github.com/iterative/cml) follows the same rules of markdown)

**Comment by Anirban166** [comment]
AFAIK there is also no strict vertical limit to the content for a GitHub comment (just 65536 characters at max), and [cml](https://github.com/iterative/cml) follows the same rules of markdown

**Comment by Anirban166** [comment]
An example from my fork: (looks the same to me)
Before table - https://github.com/Anirban166/data.table/pull/22#issuecomment-2380077011
After - https://github.com/Anirban166/data.table/pull/23#issuecomment-2380259418

So I suspect it's the image dimensions that needs to be adjusted (i.e. aside from the width fix making to CRAN, maybe height too), or made similar to tests_all_facet.png

**Comment by tdhock** [comment]
I believe the "too much compression of the image" issue that Michael mentioned is already fixed in atime on github, I just need to do a CRAN submission "(Note: Currently, the width of the preview plot is based on the total number of test cases as opposed to the number considered for generating the image, but it has been fixed and we just need to wait for that updated version of atime to be on CRAN, as that issue should automatically be resolved - Needs no updating on the GHA end)" https://github.com/Rdatatable/data.table/pull/6545#issue-2554428396

**Comment by Anirban166** [comment]
At first I thought that too, but then I saw the plots in old PRs ([e.g.](https://github.com/Rdatatable/data.table/pull/6464)) and they seem to have more height as well. But then its possible for the height to be shrunk if the width is stretched, so I guess that works

**Comment by Anirban166** [comment]
At first I thought that too, but then I saw the plots in old PRs ([e.g.](https://github.com/Rdatatable/data.table/pull/6464#issuecomment-2330023480)) and they seem to have more height as well. But then its possible for the height to be shrunk if the width is stretched, so I guess that works

**Comment by MichaelChirico** [comment]
> I just need to do a CRAN submission

Nice! I saw that comment but wasn't sure what it meant. Let's wait for a new atime & hope that resolves it.

**Pull request merged**